### ISSUE TITLE
Continuous integration with travis-ci

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#672]
[Travis CI](https://travis-ci.org/) supports continuous integration testing for open source projects.

This could create debian packages for daily builds, run unit tests, check stuff with lint, and verify that sources are compatible with both Python 2 and 3.

**Notes**

- pytest is part of the Anaconda distro.
- travis-ci requires a working build/test script.  Some possibilities:

1.  Run unitTest.leo from a bridge from, say, leo_unit_tests.py, which pytest would then run.
2. A script to convert all `@test` nodes to separate unit tests. 
   The make_stub_files project used such a script, iirc.

**Comment by maphew** [comment]
I succeeded in getting a minimal travis-ci done, in the branch **travis**. The build passes, but only because it's not testing leo ;-)  By default travis searches the project and all sub-directories and processes any file with `test` in the name. I hard coded it test only one file in `.travis.yml`

Passes:
https://travis-ci.org/leo-editor/leo-editor/builds/420717673
```
script:
    - pytest test_pass.py
```

Fails, but with a deliberate bad test file included, so successful:
https://travis-ci.org/leo-editor/leo-editor/builds/420716344
```
script:
    - pytest test_*.py
```

I've never used a continuous integration tool before. It's really neat. Each build takes about 1 minute to get to the point of being able to run the tests. I can easily see how this will help hugely in verifying and troubleshooting the pip packaging let alone the "real testing".

**Comment by maphew** [comment]
Travis branch now successfully:

- fetches named release tag from github (5.8-b1)
- builds a Leo distribution wheel
- installs the wheel
- tests Leo in console mode

It's essentially the same recipe as _LeoDist.leo: How to... >> ...build PyPi package_
It takes about 2 minutes, which is much faster than my machine. Very nice. I do feel a smidgen of guilt for committing a single character typo change which then results in all that work (!), which is done for us for free.

Log - https://travis-ci.org/leo-editor/leo-editor/builds/421909199

Next up: investigate [automatic deployment to pypi ](https://docs.travis-ci.com/user/deployment/pypi/)

----
Back to the original goal, testing: can Leo run the unit tests, or a portion of them, in headless console mode? Possibly more straightforward than converting them for pytest, assuming it's works / is feasible.

    leo-console --script=./leo/test/cmdline-unittests.leo

**Comment by edreamleo** [comment]
On Wed, Aug 29, 2018 at 1:17 AM matt wilkie <notifications@github.com>
wrote:

> Travis branch now successfully:
>
Excellent.  Thanks for this.

------------------------------
>
> Back to the original goal, testing: can Leo run the unit tests, or a
> portion of them, in headless console mode? Possibly more straightforward
> than converting them for pytest, assuming it's works / is feasible.
>
leo-console --script=./leo/test/cmdline-unittests.leo
>
> Hmm.  It's an interesting problem. Perhaps the code in leoTest.leo could
examine a new command-line arg, and do sys.exit(1) on failure. Something
like:

leo --travis-ci unitTest.leo

Edward

**Comment by maphew** [comment]
pypi deployment partway there, pending https://github.com/travis-ci/travis-ci/issues/10052

**Comment by maphew** [comment]
**pypi deployment works!** Though I had to employ some trickier to get there. Read the linked travis ticket for details. Once this work is in master branch all trickery can be dropped as as travis triggers pypi deployment on presence of a tag in the commit.

https://pypi.org/project/leo/5.8b2/#history

I opened a pull request to _devel_ in #972 to enable travis in master (on the next devel>>master merge).
